File: src/main/java/com/kamikazejam/datastore/base/storage/StorageMethods.java
package com.kamikazejam.datastore.base.storage;

import com.kamikazejam.datastore.base.Cache;
import com.kamikazejam.datastore.base.Store;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.NotNull;

import java.util.Optional;

/**
 * Methods that all storage layers must implement.
 */
@SuppressWarnings({"UnusedReturnValue", "unused"})
public interface StorageMethods<K, X extends Store<X, K>> {

    /**
     * Retrieve a Store from this store.
     */
    Optional<X> get(@NotNull K key);

    /**
     * Save a Store to this store.
     * @return If the Store was saved.
     */
    boolean save(@NotNull X store);

    /**
     * Check if a Store is stored in this store.
     */
    boolean has(@NotNull K key);

    /**
     * Check if a Store is stored in this store.
     */
    boolean has(@NotNull X store);

    /**
     * Remove a Store from this store.
     *
     * @return If the Store existed, and was removed.
     */
    boolean remove(@NotNull K key);

    /**
     * Remove a Store from this store.
     *
     * @return If the Store existed, and was removed.
     */
    boolean remove(@NotNull X store);

    /**
     * Retrieve all Stores from this store.
     */
    @NotNull
    Iterable<X> getAll();

    /**
     * Retrieve all Store keys from this store.
     */
    @NotNull
    Iterable<K> getKeys();

    /**
     * Retrieve all Store keys (in string form) from this store.
     * Uses {@link Cache#keyToString(Object)} to convert keys to strings.
     */
    @NotNull
    Iterable<String> getKeyStrings(@NotNull Cache<K, X> cache);

    /**
     * Clear all Stores from this store. No Stores are deleted, just removed from memory.
     */
    long clear();

    /**
     * Gets the name of this storage layer.
     */
    @NotNull
    String getLayerName();

    /**
     * @return How many objects are in this Store
     */
    long size();

    /**
     * @return True IFF this Store is a database (stores data elsewhere)
     */
    @ApiStatus.Internal
    boolean isDatabase();

}

File: src/main/kotlin/com/kamikazejam/datastore/base/storage/StorageMethods.kt
package com.kamikazejam.datastore.base.storage

import com.kamikazejam.datastore.base.Collection
import com.kamikazejam.datastore.base.Store
import kotlinx.coroutines.flow.Flow
import org.jetbrains.annotations.ApiStatus

/**
 * Methods that all storage layers must implement.
 */
@Suppress("unused")
interface StorageMethods<K, X : Store<X, K>> {
    /**
     * Retrieve a Store from this database.
     */
    suspend fun get(key: K): X?

    /**
     * Save a Store to this database.
     * @return If the Store was saved.
     */
    suspend fun save(store: X): Boolean

    /**
     * Check if a Store is stored in this database.
     */
    suspend fun has(key: K): Boolean

    /**
     * Check if a Store is stored in this database.
     */
    suspend fun has(store: X): Boolean

    /**
     * Remove a Store from this database.
     * @return If the Store existed, and was removed.
     */
    suspend fun remove(key: K): Boolean

    /**
     * Remove a Store from this database.
     * @return If the Store existed, and was removed.
     */
    suspend fun remove(store: X): Boolean

    /**
     * Removes all Stores from this storage. In the local case they are removed from cache, in the database case they are DELETED
     * @return How many objects were removed.
     */
    suspend fun removeAll(): Long

    /**
     * Retrieve all Stores from this database.
     */
    suspend fun getAll(): Flow<X>

    /**
     * Retrieve all Store keys from this database.
     */
    suspend fun getKeys(): Flow<K>

    /**
     * Retrieve all Store keys (in string form) from this database.
     * Uses [Collection.keyToString] to convert keys to strings.
     */
    suspend fun getKeyStrings(collection: Collection<K, X>): Flow<String>

    /**
     * Gets the name of this storage layer.
     */
    val layerName: String

    /**
     * @return How many objects are in this database
     */
    suspend fun size(): Long

    @get:ApiStatus.Internal
    val isDatabase: Boolean
}
